kmip/services/server/store/repo.py

```diff
@@ -0,0 +1,194 @@
+
+    A ManagedObjectRepo not only manages the storage of managed objects, but it
+    also must provide methods for transaction management. The methods allow a
+    caller to begin, commit, and rollback transactions. This allows a caller to
+    perform atomic transactions.
+
+    NOTE: This interface is intended for single-threaded applications. In the
+    future this interface is likely to be enhanced for multi-threaded
+    applications. When callers of the interface call get_transaction this will
+    create a class variable to store the current transaction. Then subsequent
+    calls to save, get, commit, etc. will be assumed to be in the same
+    transaction.
+
+    The future likely change will be to add a return value to get_transaction
+    that returns a reference to a transaction object. In addition each method
+    will add a kwarg for transaction.
+    """
+
+    @abc.abstractmethod
+    def begin_transaction(self):
+        """Begins a transaction
+
+        This begins a new transaction. If a transaction is already in process
+        then this throws an error.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def commit(self):
+        """Commits the transaction
+
+        This commits the transaction and ends the current transaction. After
+        this is called then begin_transaction may be called again to begin
+        another transaction.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def rollback(self):
+        """Rollback the current transaction
+
+        The rolls back the current transaction and ends the current
+        transaction. After this is called then begin_transaction may be called
+        again to begin another transaction.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def save(self, managed_object):
+        """Save a managed object
+
+        This saves a managed object into the repository and returns a UID
+        string that can be used to reference the object in the repository.
+        :param managed_object: managed object to save
+        :returns: a UID string that can be used to retrieve the object later
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get(self, uid):
+        """Retrieve a managed object
+
+        Retrieve a managed object from the repository. The UID is used to
+        identify the managed object to return. The UID is returned from the
+        save call.
+
+        :param uuid: UUID of the managed object
+        :returns: managed_object if object exists, otherwise None
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def search(self, attributes, max_items,
+               storage_status_mask=(enums.StorageStatusMask.ONLINE_STORAGE),
+               object_group_member=None):
+        """Search for a managed object.
+
+        Search for and return managed objects that have all of the attributes.
```

Comment:
The locate operation only returns the UIDs of the managed objects. Do you want to just return UIDs or do you want the entire object? I was thinking that if I returned everything then the server can iterate over the results and pull out the UIDs if that is all it wants.